content/en/blog/_posts/2022-12-05-forensic-container-checkpointing/index.md

```diff
@@ -0,0 +1,205 @@
+
+```shell
+curl -X POST "https://localhost:10250/checkpoint/default/counters/counter"
+```
+
+For completeness the following `curl` command-line options are necessary to
+have `curl` accept the *kubelet*'s self signed certificate and authorize the
+use of the *kubelet* `checkpoint` API:
+
+```shell
+--insecure --cert /var/run/kubernetes/client-admin.crt --key /var/run/kubernetes/client-admin.key
+```
+
+Triggering this **kubelet** API will request the creation of a checkpoint from
+CRI-O. CRI-O requests a checkpoint from your low-level runtime (for example,
+`runc`). Seeing that request, `runc` invokes the `criu` tool
+to do the actual checkpointing.
+
+Once the checkpointing has finished the checkpoint should be available at
+`/var/lib/kubelet/checkpoints/checkpoint-<pod-name>_<namespace-name>-<container-name>-<timestamp>.tar`
+
+You could then use that tar archive to restore the container somewhere else.
+
+### Restore a checkpointed container outside of Kubernetes (with CRI-O) {#restore-checkpointed-container-standalone}
+
+With the checkpoint tar archive it is possible to restore the container outside
+of Kubernetes in a sandboxed instance of CRI-O. For better user experience
+during restore, I recommend that you use the latest version of CRI-O from the
+*main* CRI-O GitHub branch. If you're using CRI-O v1.25, you'll need to
+manually create certain directories Kubernetes would create before starting the
+container.
+
+The first step to restore a container outside of Kubernetes is to create a pod sandbox
+using *crictl*:
+```shell
+crictl runp pod-config.json
+```
+
+Then you can restore the previously checkpointed container into the newly created pod sandbox:
+```shell
+crictl create <POD_ID> container-config.json pod-config.json
+```
+
+Instead of specifying a container image in a registry in `container-config.json`
+you'd need specify the path to the checkpoint archive that you created earlier:
+```json
+{
+  "metadata": {
+      "name": "counter"
+  },
+  "image":{
+      "image": "/var/lib/kubelet/checkpoints/<checkpoint-archive>.tar"
+  }
+}
+```
+
+Next, run `crictl start <CONTAINER_ID>` to start that container, and then a
+copy of the previously checkpointed container should be running.
+
+### Restore a checkpointed container within of Kubernetes {#restore-checkpointed-container-k8s}
+
+To restore the previously checkpointed container directly in Kubernetes it is
+necessary to convert the checkpoint archive into an image that can be pushed to
+a registry.
+
+One possible way to convert the local checkpoint archive consists of the
+following steps with the help of [buildah](https://buildah.io/):
+```shell
+newcontainer=$(buildah from scratch)
+buildah add $newcontainer /var/lib/kubelet/checkpoints/checkpoint-<pod-name>_<namespace-name>-<container-name>-<timestamp>.tar /
+buildah config --annotation=io.kubernetes.cri-o.annotations.checkpoint.name=<container-name> $newcontainer
+buildah commit $newcontainer checkpoint-image:latest
```

Comment:
The following line is necessary to clean up the `working-container` used by buildah after `buildah commit`. It would be good to add it in the example as well.
```shell
buildah rm working-container
````

Reply:
done